### ISSUE TITLE
Pyright thinks that a global variable being shadowed is unbounded.

### ISSUE THREAD

**Issue by jmg-duarte** [microsoft/pyright#3076]
**Describe the bug**
Pyright thinks that a global variable (which is very much declared) is unbounded if I try to replace it in a call like `a = f(a)`.

**To Reproduce**
```python
import logging
from . import ContextAdapter # Doesn't matter which Adapter it is

logger = logging.getLogger()

def f():
    # Pyright thinks that the `logger` inside the call is unbounded
    logger = ContextAdapter(logger, {"shadow": "failed"}) 
```

**Expected behavior**
Pyright not raising any issues.

**VS Code extension or command-line**
VSCode extension.

**Comment by jmg-duarte** [comment]
The Python documentation seems to (kinda) disagree.

> In Python, variables that are only referenced inside a function are implicitly global. If a variable is assigned a value anywhere within the function’s body, it’s assumed to be a local unless explicitly declared as global.

https://docs.python.org/3/faq/programming.html?highlight=global%20keyword#what-are-the-rules-for-local-and-global-variables-in-python

And in my case, that is precisely what I am going for. I do not want to mutate the global variable, I want to declare a local one that shadows the global.

Pyright should follow Python's behavior, read the right `logger` as global and the left one as local. 
Maybe I am missing something but telling me the variable is unbounded when it clearly isn't seems to go against the language.

**Comment by JelleZijlstra** [comment]
Your code doesn't work at runtime.

```
Python 3.11.0a5+ (heads/main:6f1efd19a7, Feb 16 2022, 17:43:30) [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> 
>>> def ContextAdapter(*args, **kwargs):
...     return 1
... 
>>> logger = logging.getLogger()
>>> 
>>> def f():
...     # Pyright thinks that the `logger` inside the call is unbounded
...     logger = ContextAdapter(logger, {"shadow": "failed"}) 
... 
>>> f()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in f
UnboundLocalError: cannot access local variable 'logger' where it is not associated with a value
```
